File: Week-Wise-Assignments/Assignment 3/Compressed Solutions/Python Programs/6_odd_even_nodes.py
class Node:
    def __init__(self, d): self.data, self.next = d, None

class LL:
    def __init__(self): self.head = None
    def add(self, d): n = Node(d); (setattr(self, 'head', n) if not self.head else [t for t in [self.head] if not [setattr(t, 'next', n) for t in iter(lambda x=self.head: x.next if x.next else None, None)]])
    def printOE(self):
        odd, even, t = [], [], self.head
        while t: (odd if t.data % 2 else even).append(t.data); t = t.next
        print(f"Odd: {' -> '.join(map(str, odd)) + ' -> None' if odd else 'None'}")
        print(f"Even: {' -> '.join(map(str, even)) + ' -> None' if even else 'None'}")

ll = LL()
[ll.add(i) for i in [12, 7, 18, 3, 24, 9, 15, 6]]
ll.printOE()

File: Week-Wise-Assignments/Assignment 3/Compressed Solutions/C Programs/6_odd_even_nodes.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node { int data; struct Node* next; } Node;
Node* cn(int d) { Node* n = malloc(sizeof(Node)); n->data = d; n->next = NULL; return n; }
Node* ial(Node* h, int d) { Node* n = cn(d); if (!h) return n; Node* t = h; while (t->next) t = t->next; t->next = n; return h; }
void printOE(Node* h) {
    printf("Odd: "); Node* t = h; while (t) { if (t->data % 2) printf("%d ", t->data); t = t->next; } printf("\n");
    printf("Even: "); t = h; while (t) { if (!(t->data % 2)) printf("%d ", t->data); t = t->next; } printf("\n");
}
int main() {
    Node* h = NULL; int vals[] = {12, 7, 18, 3, 24, 9, 15, 6};
    for (int i = 0; i < 8; i++) h = ial(h, vals[i]);
    printOE(h);
    return 0;
}